lib/elixir/lib/string.ex

```diff
@@ -618,6 +618,10 @@ defmodule String do
   @spec replace_leading(t, t, t) :: t
   def replace_leading(string, match, replacement)
       when is_binary(string) and is_binary(match) and is_binary(replacement) do
+    if match == "" do
```

Comment:
yes. using if saves us from two `is_binary` guard checks (one for `string`, one for `replacement`),
if you want to use guards, which seems reasonable, is that you need to create a private function, check only one in the public function, and not check anything in the private one.

```elixir
def replace_leading(string, match, replacement) when is_binary(string) and is_binary(match) and is_binary(replacement),
  do: do_replace_leading(string, match, replacement)

defp replace_leading(_string, "", _replacement) do
  raise ArgumentError, "cannot use an empty string as the match to replace"
end

defp replace_leading(string, match, replacement) do
  # implementation
end
```